File: Java/Graphs/24_JobScheduling.java
class Solution
{
    //Function to find the maximum profit and the number of jobs done.
    int[] JobScheduling(Job arr[], int n)
    {
        int[] ans = new int[2];
        Arrays.sort(arr,(a,b)->{
            return b.profit - a.profit;
        });
        int maxDeadline = 0;
        for(int i=0;i<n;i++){
            maxDeadline = Math.max(maxDeadline,arr[i].deadline);
        }
        boolean[] vis = new boolean[maxDeadline+1];
        int maxJobs = 0;
        int maxProfit = 0;
        for(Job a: arr){           
            for(int i=a.deadline;i>=1;i--){
                if(!vis[i]){
                    vis[i] = true;
                    maxJobs++;
                    maxProfit += a.profit;
                    break;
                }
            }
        }
        ans[0] = maxJobs;
        ans[1] = maxProfit;
        return ans;
    }
}

File: C++/Graphs/24_JobScheduling.cpp
int find(int x, vector<int> &parent)
{
    if (parent[x] == x)
        return parent[x] = x;
    return parent[x] = find(parent[x], parent);
}

static bool cmp(Job a, Job b)
{
    return a.profit > b.profit;
}
vector<int> JobScheduling(Job arr[], int n)
{
    // your code here
    int jobs = 0;
    int maxProfit = 0;
    int maxDeadline = 0;
    // get the maxDeadline
    for (int i = 0; i < n; i++)
    {
        maxDeadline = max(arr[i].dead, maxDeadline);
    }
    // initialise parent array
    vector<int> parent(maxDeadline + 1, 0);
    for (int i = 0; i <= maxDeadline; i++)
        parent[i] = i;

    sort(arr, arr + n, cmp);

    for (int i = 0; i < n; i++)
    {
        int slot = find(arr[i].dead, parent);

        if (slot > 0)
        {
            int slot2 = find(slot - 1, parent);
            parent[slot] = slot2;
            jobs++;
            maxProfit += arr[i].profit;
        }
    }

    return {jobs, maxProfit};
}